### PR TITLE
fix(compiler): always apply bindings on elements before host bindings…

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: modules/@angular/compiler/src/view_compiler/view_compiler.ts [hunk 1]
```diff
@@ -288,24 +288,24 @@ class ViewBuilder implements TemplateAstVisitor, LocalResolver, BuiltinConverter
...
-    const {flags, usedEvents, queryMatchesExpr, hostBindings, hostEvents} =
+    const {flags, usedEvents, queryMatchesExpr, hostBindings: dirHostBindings, hostEvents} =
         this._visitElementOrTemplate(nodeIndex, ast);
 
     let inputDefs: o.Expression[] = [];
     let outputDefs: o.Expression[] = [];
     if (elName) {
-      ast.inputs.forEach(
-          (inputAst) => hostBindings.push({context: COMP_VAR, value: inputAst.value}));
+      const hostBindings = ast.inputs.map((inputAst) => {
```
Comment (vicb) [inline modules/@angular/compiler/src/view_compiler/view_compiler.ts]: nit: 
```
const hostBindings = ast.inputs.map((inputAst) => ({
  context: ...
}).concat(dirHostBinding);

Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/view_compiler.ts]: Done

File: modules/@angular/compiler/src/view_compiler/view_compiler.ts [hunk 2]
```diff
@@ -517,6 +517,7 @@ class ViewBuilder implements TemplateAstVisitor, LocalResolver, BuiltinConverter
       return {
         value: (<ASTWithSource>hostBindingAst.value).ast,
         context: dirContextExpr,
+        bindingDef: elementBindingDef(hostBindingAst, dirAst)
```
Comment (vicb) [inline modules/@angular/compiler/src/view_compiler/view_compiler.ts]: nit: trailing ,

File: modules/@angular/core/test/linker/change_detection_integration_spec.ts [hunk 3]
```diff
@@ -1241,6 +1243,34 @@ export function main() {
...
     });
+
+    describe('class binding', () => {
+      it('should coordinate class attribute and class host binding', () => {
+        @Component({template: `<div class="{{initClasses}}" someDir></div>`})
+        class Comp {
+          initClasses = 'init';
+        }
+
+        @Directive({
```
Comment (vicb) [inline modules/@angular/core/test/linker/change_detection_integration_spec.ts]: nit: 1 line

File: modules/@angular/compiler/src/view_compiler/view_compiler.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/view_compiler.ts]: Done

File: modules/@angular/core/test/linker/change_detection_integration_spec.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/core/test/linker/change_detection_integration_spec.ts]: Done

File: modules/@angular/core/test/linker/change_detection_integration_spec.ts [hunk 4]
```diff
@@ -1241,6 +1243,34 @@ export function main() {
...
+          selector: '[someDir]',
+        })
+        class SomeDir {
+          @HostBinding('class.foo')
+          fooClass = true;
+        }
+
+        TestBed.configureCompiler(
```
Comment (vicb) [inline modules/@angular/core/test/linker/change_detection_integration_spec.ts]:
```
const ctx = TestBed.configureCompiler(...)
   .configureTestingModule(...)
   .createComponent(...);
```

File: modules/@angular/core/test/linker/change_detection_integration_spec.ts [hunk 5]
```diff
@@ -1284,14 +1314,15 @@ function patchLoggingRendererV2(rendererFactory: RendererFactoryV2, log: RenderL
     const origSetProperty = renderer.setProperty;
     const origSetValue = renderer.setValue;
     renderer.setProperty = function(el: any, name: string, value: any): void {
+      console.log('>>>', el, name, value);
```
Comment (vicb) [inline modules/@angular/core/test/linker/change_detection_integration_spec.ts]: revert

File: modules/@angular/core/test/linker/change_detection_integration_spec.ts (hunk 4 above)
Comment (tbosch) [inline modules/@angular/core/test/linker/change_detection_integration_spec.ts]: Done

**Review by vicb — APPROVED** [review]

File: modules/@angular/core/test/linker/change_detection_integration_spec.ts (hunk 5 above)
Comment (tbosch) [inline modules/@angular/core/test/linker/change_detection_integration_spec.ts]: Done

File: modules/@angular/compiler/src/view_compiler/view_compiler.ts [hunk 6]
```diff
@@ -288,24 +288,22 @@ class ViewBuilder implements TemplateAstVisitor, LocalResolver, BuiltinConverter
...
     let inputDefs: o.Expression[] = [];
     let outputDefs: o.Expression[] = [];
     if (elName) {
-      ast.inputs.forEach(
-          (inputAst) => hostBindings.push({context: COMP_VAR, value: inputAst.value}));
+      const hostBindings = ast.inputs.map((inputAst) => ({
+                                            context: COMP_VAR as o.Expression,
+                                            value: inputAst.value,
+                                            bindingDef: elementBindingDef(inputAst, null),
+                                          }));
```
Comment (vicb) [inline modules/@angular/compiler/src/view_compiler/view_compiler.ts]: .concat ?

Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/view_compiler.ts]: Done

**Pull request merged**